File: services/chair.service.js
/**
 * BarberEase – Chair Service
 * 
 * Business logic for chair management:
 *   - Get all chairs with status
 *   - Update chair status
 *   - Get available chairs
 *   - Get chair statistics
 *   - Reserve a chair (with timeout)
 *   - Release expired reservations
 */

const { Op } = require('sequelize');
const { Chair, Customer } = require('../models');
const config = require('../config/app');

/**
 * Get all chairs with their current status.
 */
async function getAllChairs() {
  return await Chair.findAll({
    where: { is_active: true },
    include: [
      { model: Customer, as: 'reservedByCustomer', attributes: ['id', 'name'] }
    ],
    order: [['chair_number', 'ASC']]
  });
}

/**
 * Get chair by ID.
 */
async function getChairById(id) {
  const chair = await Chair.findByPk(id, {
    include: [
      { model: Customer, as: 'reservedByCustomer', attributes: ['id', 'name'] }
    ]
  });

  if (!chair) {
    const error = new Error('Chair not found');
    error.statusCode = 404;
    throw error;
  }

  return chair;
}

/**
 * Get only available chairs.
 */
async function getAvailableChairs() {
  return await Chair.findAll({
    where: { status: 'available', is_active: true },
    order: [['chair_number', 'ASC']]
  });
}

/**
 * Update chair status (admin operation).
 * Valid transitions:
 *   any → maintenance (admin override)
 *   maintenance → available
 */
async function updateChairStatus(chairId, newStatus) {
  const chair = await Chair.findByPk(chairId);

  if (!chair) {
    const error = new Error('Chair not found');
    error.statusCode = 404;
    throw error;
  }

  const updateData = { status: newStatus };

  // Clear reservation data when releasing
  if (newStatus === 'available') {
    updateData.reserved_by = null;
    updateData.reserved_at = null;
  }

  await chair.update(updateData);
  return chair;
}

/**
 * Create a new chair (admin operation).
 */
async function createChair(data) {
  return await Chair.create({
    chair_number: data.chair_number,
    name: data.name || `Chair ${data.chair_number}`,
    status: 'available',
    is_active: true
  });
}

/**
 * Get chair statistics summary.
 * Returns counts by status for dashboard display.
 */
async function getChairStats() {
  const chairs = await Chair.findAll({ where: { is_active: true } });

  const stats = {
    total: chairs.length,
    available: 0,
    reserved: 0,
    occupied: 0,
    maintenance: 0
  };

  chairs.forEach(chair => {
    if (stats[chair.status] !== undefined) {
      stats[chair.status]++;
    }
  });

  return stats;
}

/**
 * Release chairs whose reservation has expired.
 * Called periodically to auto-release reservations
 * that exceed the configured timeout (default 5 min).
 */
async function releaseExpiredReservations() {
  const timeoutMs = config.reservationTimeoutMs;
  const cutoff = new Date(Date.now() - timeoutMs);

  const expired = await Chair.findAll({
    where: {
      status: 'reserved',
      reserved_at: { [Op.lt]: cutoff }
    }
  });

  for (const chair of expired) {
    await chair.update({
      status: 'available',
      reserved_by: null,
      reserved_at: null
    });
  }

  return expired.length;
}

/**
 * Temporarily reserve a chair for 5 minutes.
 */
async function reserveChair(chairId, customerId) {
  const chair = await Chair.findByPk(chairId);

  if (!chair) {
    const error = new Error('Chair not found');
    error.statusCode = 404;
    throw error;
  }

  if (chair.status !== 'available' || !chair.is_active) {
    const error = new Error('Chair is no longer available');
    error.statusCode = 409;
    throw error;
  }

  await chair.update({
    status: 'reserved',
    reserved_by: customerId,
    reserved_at: new Date()
  });

  return chair;
}

/**
 * Release a temporarily reserved chair.
 */
async function releaseChair(chairId, customerId) {
  const chair = await Chair.findByPk(chairId);

  if (!chair) {
    const error = new Error('Chair not found');
    error.statusCode = 404;
    throw error;
  }

  // Only release if it was reserved by this customer
  if (chair.status === 'reserved' && chair.reserved_by === customerId) {
    await chair.update({
      status: 'available',
      reserved_by: null,
      reserved_at: null
    });
  }

  return chair;
}

module.exports = {
  getAllChairs,
  getChairById,
  getAvailableChairs,
  updateChairStatus,
  createChair,
  getChairStats,
  releaseExpiredReservations,
  reserveChair,
  releaseChair
};

File: src/main/java/com/barberease/services/ChairService.java
package com.barberease.services;

import com.barberease.models.Chair;
import com.barberease.models.Customer;
import com.barberease.repositories.ChairRepository;
import com.barberease.repositories.CustomerRepository;
import com.barberease.websocket.BarberWebSocketHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import com.barberease.repositories.AppointmentRepository;
import com.barberease.repositories.QueueRepository;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@Transactional
public class ChairService {

    @Autowired
    private ChairRepository chairRepository;

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private AppointmentRepository appointmentRepository;

    @Autowired
    private QueueRepository queueRepository;

    @Autowired
    private BarberWebSocketHandler webSocketHandler;

    @Value("${app.reservation-timeout-ms}")
    private long reservationTimeoutMs;

    public void reconcileChairStatuses() {
        List<Chair> chairs = chairRepository.findAll();
        boolean changed = false;
        for (Chair chair : chairs) {
            if ("occupied".equals(chair.getStatus())) {
                boolean hasActiveAppt = appointmentRepository.findByChairId(chair.getId()).stream()
                        .anyMatch(a -> "in_progress".equals(a.getStatus()));
                boolean hasActiveQueue = queueRepository.findByChairIdAndStatusInOrderByPositionAsc(
                        chair.getId(), Arrays.asList("serving", "called")).stream()
                        .findAny().isPresent();

                if (!hasActiveAppt && !hasActiveQueue) {
                    chair.setStatus("available");
                    chair.setReservedByCustomer(null);
                    chair.setReservedAt(null);
                    chairRepository.save(chair);
                    changed = true;
                }
            }
        }
        if (changed) {
            webSocketHandler.broadcast("chair-update");
        }
    }

    public List<Chair> getAllChairs() {
        reconcileChairStatuses();
        return chairRepository.findByIsActiveTrueOrderByChairNumberAsc();
    }

    public Chair getChairById(Long id) {
        return chairRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Chair not found"));
    }

    public List<Chair> getAvailableChairs() {
        reconcileChairStatuses();
        return chairRepository.findByStatusAndIsActiveTrueOrderByChairNumberAsc("available");
    }

    public Chair updateChairStatus(Long chairId, String newStatus) {
        Chair chair = getChairById(chairId);

        chair.setStatus(newStatus);
        if ("available".equals(newStatus)) {
            chair.setReservedByCustomer(null);
            chair.setReservedAt(null);
        }

        return chairRepository.save(chair);
    }

    public Chair createChair(Chair data) {
        data.setStatus("available");
        data.setActive(true);
        if (data.getName() == null) {
            data.setName("Chair " + data.getChairNumber());
        }
        return chairRepository.save(data);
    }

    public Map<String, Integer> getChairStats() {
        List<Chair> chairs = getAllChairs();

        Map<String, Integer> stats = new HashMap<>();
        stats.put("total", chairs.size());
        stats.put("available", 0);
        stats.put("reserved", 0);
        stats.put("occupied", 0);
        stats.put("maintenance", 0);

        for (Chair chair : chairs) {
            String status = chair.getStatus();
            stats.put(status, stats.getOrDefault(status, 0) + 1);
        }

        return stats;
    }

    @Scheduled(fixedRate = 15000)
    public void cleanExpiredReservations() {
        LocalDateTime cutoff = LocalDateTime.now().minusNanos(reservationTimeoutMs * 1_000_000);
        List<Chair> expired = chairRepository.findByStatusAndReservedAtBefore("reserved", cutoff);

        if (!expired.isEmpty()) {
            for (Chair chair : expired) {
                chair.setStatus("available");
                chair.setReservedByCustomer(null);
                chair.setReservedAt(null);
                chairRepository.save(chair);
            }
            System.out.println("⏰ Cleaned up " + expired.size() + " expired chair reservations.");
            webSocketHandler.broadcast("chair-update");
        }

        reconcileChairStatuses();
    }

    public Chair reserveChair(Long chairId, Long customerId) {
        Chair chair = getChairById(chairId);
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new IllegalArgumentException("Customer not found"));

        if (!chair.isActive() || "maintenance".equals(chair.getStatus())) {
            throw new IllegalStateException("Chair station is currently under maintenance.");
        }

        chair.setReservedByCustomer(customer);
        chair.setReservedAt(LocalDateTime.now());

        return chairRepository.save(chair);
    }

    public Chair releaseChair(Long chairId, Long customerId) {
        Chair chair = getChairById(chairId);

        if ("reserved".equals(chair.getStatus()) && 
                chair.getReservedByCustomer() != null && 
                chair.getReservedByCustomer().getId().equals(customerId)) {
            chair.setStatus("available");
            chair.setReservedByCustomer(null);
            chair.setReservedAt(null);
            return chairRepository.save(chair);
        }

        return chair;
    }
}
